File: stripe/src/test/java/com/stripe/android/view/PaymentMethodsActivityTest.java
package com.stripe.android.view;

import android.content.Context;
import android.content.Intent;
import android.view.View;
import android.widget.ProgressBar;

import androidx.recyclerview.widget.RecyclerView;
import androidx.test.core.app.ApplicationProvider;

import com.stripe.android.ApiKeyFixtures;
import com.stripe.android.CustomerSession;
import com.stripe.android.CustomerSessionTestHelper;
import com.stripe.android.PaymentConfiguration;
import com.stripe.android.R;
import com.stripe.android.model.PaymentMethod;
import com.stripe.android.model.PaymentMethodFixtures;

import java.util.Objects;

import org.junit.After;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.ArgumentCaptor;
import org.mockito.Captor;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;
import org.robolectric.RobolectricTestRunner;
import org.robolectric.Shadows;
import org.robolectric.shadows.ShadowActivity;

import static android.app.Activity.RESULT_OK;
import static com.stripe.android.PaymentSession.EXTRA_PAYMENT_SESSION_ACTIVE;
import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertNotNull;
import static org.junit.Assert.assertTrue;
import static org.mockito.ArgumentMatchers.eq;
import static org.mockito.Mockito.reset;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;


/**
 * Test class for {@link PaymentMethodsActivity}.
 */
@RunWith(RobolectricTestRunner.class)
public class PaymentMethodsActivityTest extends BaseViewTest<PaymentMethodsActivity> {
    @Mock private CustomerSession mCustomerSession;
    @Captor private ArgumentCaptor<CustomerSession.PaymentMethodsRetrievalListener> mListenerArgumentCaptor;

    private Context mContext;
    private PaymentMethodsActivity mPaymentMethodsActivity;
    private ProgressBar mProgressBar;
    private RecyclerView mRecyclerView;
    private View mAddCardView;
    private ShadowActivity mShadowActivity;

    public PaymentMethodsActivityTest() {
        super(PaymentMethodsActivity.class);
    }

    @Before
    public void setup() {
        MockitoAnnotations.initMocks(this);

        mContext = ApplicationProvider.getApplicationContext();

        CustomerSessionTestHelper.setInstance(mCustomerSession);
        PaymentConfiguration.init(mContext, ApiKeyFixtures.FAKE_PUBLISHABLE_KEY);

        mPaymentMethodsActivity = createActivity(
                new PaymentMethodsActivityStarter.Args.Builder()
                        .setPaymentConfiguration(PaymentConfiguration.getInstance(mContext))
                        .build()
        );
        mShadowActivity = Shadows.shadowOf(mPaymentMethodsActivity);

        mProgressBar = mPaymentMethodsActivity.findViewById(R.id.payment_methods_progress_bar);
        mRecyclerView = mPaymentMethodsActivity.findViewById(R.id.payment_methods_recycler);
        mAddCardView = mPaymentMethodsActivity
                .findViewById(R.id.payment_methods_add_card);
    }

    @After
    @Override
    public void tearDown() {
        super.tearDown();
    }

    @Test
    public void onCreate_callsApiAndDisplaysProgressBarWhileWaiting() {
        assertNotNull(mProgressBar);
        assertNotNull(mRecyclerView);
        assertNotNull(mAddCardView);

        verify(mCustomerSession).getPaymentMethods(eq(PaymentMethod.Type.Card),
                mListenerArgumentCaptor.capture());

        assertEquals(View.VISIBLE, mProgressBar.getVisibility());
        assertEquals(View.VISIBLE, mAddCardView.getVisibility());
        assertEquals(View.VISIBLE, mRecyclerView.getVisibility());

        final CustomerSession.PaymentMethodsRetrievalListener listener =
                mListenerArgumentCaptor.getValue();
        assertNotNull(listener);

        listener.onPaymentMethodsRetrieved(PaymentMethodFixtures.CARD_PAYMENT_METHODS);

        assertEquals(View.GONE, mProgressBar.getVisibility());
    }

    @Test
    public void onCreate_initialGivenPaymentMethodIsSelected() {
        // reset the mock because the activity is being re-created again
        reset(mCustomerSession);
        final PaymentMethod paymentMethod = PaymentMethodFixtures.CARD_PAYMENT_METHODS.get(0);
        assertNotNull(paymentMethod);
        mPaymentMethodsActivity = createActivity(new PaymentMethodsActivityStarter.Args.Builder()
                .setInitialPaymentMethodId(paymentMethod.id)
                .build());
        mRecyclerView = mPaymentMethodsActivity.findViewById(R.id.payment_methods_recycler);

        verify(mCustomerSession).getPaymentMethods(eq(PaymentMethod.Type.Card),
                mListenerArgumentCaptor.capture());

        final CustomerSession.PaymentMethodsRetrievalListener listener =
                mListenerArgumentCaptor.getValue();
        assertNotNull(listener);

        listener.onPaymentMethodsRetrieved(PaymentMethodFixtures.CARD_PAYMENT_METHODS);

        final PaymentMethodsAdapter paymentMethodsAdapter =
                (PaymentMethodsAdapter) mRecyclerView.getAdapter();
        assertNotNull(paymentMethodsAdapter);
        assertNotNull(paymentMethodsAdapter.getSelectedPaymentMethod());
        assertEquals(paymentMethod.id, paymentMethodsAdapter.getSelectedPaymentMethod().id);
    }

    @Test
    public void onClickAddSourceView_withoutPaymentSession_launchesAddSourceActivityWithoutLog() {
        mAddCardView.performClick();
        final ShadowActivity.IntentForResult intentForResult =
                mShadowActivity.getNextStartedActivityForResult();
        assertNotNull(intentForResult);
        assertNotNull(intentForResult.intent.getComponent());
        assertEquals(AddPaymentMethodActivity.class.getName(),
                intentForResult.intent.getComponent().getClassName());
        assertFalse(intentForResult.intent.hasExtra(EXTRA_PAYMENT_SESSION_ACTIVE));
    }

    @Test
    public void onClickAddSourceView_whenStartedFromPaymentSession_launchesActivityWithLog() {
        mPaymentMethodsActivity = createActivity(new PaymentMethodsActivityStarter.Args.Builder()
                .setPaymentConfiguration(PaymentConfiguration.getInstance(mContext))
                .setIsPaymentSessionActive(true)
                .build());
        mShadowActivity = Shadows.shadowOf(mPaymentMethodsActivity);
        mAddCardView =
                mPaymentMethodsActivity.findViewById(R.id.payment_methods_add_card);
        mAddCardView.performClick();
        final ShadowActivity.IntentForResult intentForResult =
                mShadowActivity.getNextStartedActivityForResult();
        assertNotNull(intentForResult);
        assertNotNull(intentForResult.intent.getComponent());
        assertEquals(AddPaymentMethodActivity.class.getName(),
                intentForResult.intent.getComponent().getClassName());
        assertTrue(AddPaymentMethodActivityStarter.Args.create(intentForResult.intent)
                .isPaymentSessionActive);
    }

    @Test
    public void onActivityResult_withValidPaymentMethod_refreshesPaymentMethods() {
        final PaymentMethod paymentMethod = PaymentMethodFixtures.CARD_PAYMENT_METHODS.get(2);
        assertNotNull(paymentMethod);

        final Intent resultIntent = new Intent()
                .putExtra(AddPaymentMethodActivity.EXTRA_NEW_PAYMENT_METHOD, paymentMethod);

        mPaymentMethodsActivity.onActivityResult(
                AddPaymentMethodActivityStarter.REQUEST_CODE, RESULT_OK, resultIntent
        );
        assertEquals(View.VISIBLE, mProgressBar.getVisibility());
        verify(mCustomerSession, times(2)).getPaymentMethods(
                eq(PaymentMethod.Type.Card), mListenerArgumentCaptor.capture());

        final CustomerSession.PaymentMethodsRetrievalListener listener =
                mListenerArgumentCaptor.getValue();
        assertNotNull(listener);

        listener.onPaymentMethodsRetrieved(PaymentMethodFixtures.CARD_PAYMENT_METHODS);
        assertEquals(View.GONE, mProgressBar.getVisibility());
        assertNotNull(mRecyclerView.getAdapter());
        assertEquals(4, mRecyclerView.getAdapter().getItemCount());

        final PaymentMethodsAdapter paymentMethodsAdapter =
                (PaymentMethodsAdapter) mRecyclerView.getAdapter();
        assertNotNull(paymentMethodsAdapter);
        assertNotNull(paymentMethodsAdapter.getSelectedPaymentMethod());
        assertEquals(paymentMethod.id, paymentMethodsAdapter.getSelectedPaymentMethod().id);
    }

    @Test
    public void setSelectionAndFinish_finishedWithExpectedResult() {
        verify(mCustomerSession).getPaymentMethods(eq(PaymentMethod.Type.Card),
                mListenerArgumentCaptor.capture());

        assertEquals(View.VISIBLE, mProgressBar.getVisibility());
        assertEquals(View.VISIBLE, mAddCardView.getVisibility());
        assertEquals(View.VISIBLE, mRecyclerView.getVisibility());

        final CustomerSession.PaymentMethodsRetrievalListener listener =
                mListenerArgumentCaptor.getValue();
        assertNotNull(listener);

        listener.onPaymentMethodsRetrieved(PaymentMethodFixtures.CARD_PAYMENT_METHODS);
        final PaymentMethodsAdapter paymentMethodsAdapter =
                (PaymentMethodsAdapter) mRecyclerView.getAdapter();
        assertNotNull(paymentMethodsAdapter);
        paymentMethodsAdapter.setSelectedIndex(0);

        mPaymentMethodsActivity.setSelectionAndFinish(
                PaymentMethodFixtures.CARD_PAYMENT_METHODS.get(0)
        );

        // Now it should be gone.
        assertEquals(View.GONE, mProgressBar.getVisibility());
        assertTrue(mPaymentMethodsActivity.isFinishing());
        assertEquals(RESULT_OK, mShadowActivity.getResultCode());
        final Intent intent = mShadowActivity.getResultIntent();
        assertNotNull(intent);

        final PaymentMethodsActivityStarter.Result result =
                Objects.requireNonNull(PaymentMethodsActivityStarter.Result.fromIntent(intent));
        assertEquals(PaymentMethodFixtures.CARD_PAYMENT_METHODS.get(0), result.paymentMethod);
    }
}

File: stripe/src/test/java/com/stripe/android/view/PaymentMethodsActivityTest.kt
package com.stripe.android.view

import android.app.Activity.RESULT_CANCELED
import android.app.Activity.RESULT_OK
import android.content.Context
import android.content.Intent
import android.view.View
import androidx.test.core.app.ApplicationProvider
import com.nhaarman.mockitokotlin2.KArgumentCaptor
import com.nhaarman.mockitokotlin2.argumentCaptor
import com.nhaarman.mockitokotlin2.eq
import com.nhaarman.mockitokotlin2.isNull
import com.nhaarman.mockitokotlin2.mock
import com.nhaarman.mockitokotlin2.verify
import com.stripe.android.ApiKeyFixtures
import com.stripe.android.CustomerSession
import com.stripe.android.PaymentConfiguration
import com.stripe.android.R
import com.stripe.android.model.PaymentMethod
import com.stripe.android.model.PaymentMethodFixtures
import org.junit.runner.RunWith
import org.mockito.Mockito.times
import org.robolectric.RobolectricTestRunner
import org.robolectric.Shadows.shadowOf
import kotlin.test.BeforeTest
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertNotNull
import kotlin.test.assertTrue

/**
 * Test class for [PaymentMethodsActivity].
 */
@RunWith(RobolectricTestRunner::class)
class PaymentMethodsActivityTest {
    private val customerSession: CustomerSession = mock()

    private val listenerArgumentCaptor: KArgumentCaptor<CustomerSession.PaymentMethodsRetrievalListener> = argumentCaptor()

    private val context = ApplicationProvider.getApplicationContext<Context>()
    private val activityScenarioFactory = ActivityScenarioFactory(context)

    @BeforeTest
    fun setup() {
        CustomerSession.instance = customerSession
        PaymentConfiguration.init(context, ApiKeyFixtures.FAKE_PUBLISHABLE_KEY)
    }

    @Test
    fun onCreate_callsApiAndDisplaysProgressBarWhileWaiting() {
        activityScenarioFactory.create<PaymentMethodsActivity>(
            PaymentMethodsActivityStarter.Args.Builder()
                .setPaymentConfiguration(PaymentConfiguration.getInstance(context))
                .build()
        ).use { activityScenario ->
            activityScenario.onActivity {
                val progressBar = it.viewBinding.progressBar
                val recyclerView = it.viewBinding.recycler
                val addCardView: View = it.findViewById(R.id.stripe_payment_methods_add_card)

                verify(customerSession).getPaymentMethods(
                    eq(PaymentMethod.Type.Card),
                    isNull(),
                    isNull(),
                    isNull(),
                    eq(setOf(PaymentMethodsActivity.PRODUCT_TOKEN)),
                    listenerArgumentCaptor.capture()
                )

                assertEquals(View.VISIBLE, progressBar.visibility)
                assertEquals(View.VISIBLE, addCardView.visibility)
                assertEquals(View.VISIBLE, recyclerView.visibility)

                listenerArgumentCaptor.firstValue
                    .onPaymentMethodsRetrieved(PaymentMethodFixtures.CARD_PAYMENT_METHODS)

                assertEquals(View.GONE, progressBar.visibility)
            }
        }
    }

    @Test
    fun onCreate_initialGivenPaymentMethodIsSelected() {
        val paymentMethod = PaymentMethodFixtures.CARD_PAYMENT_METHODS[0]

        activityScenarioFactory.create<PaymentMethodsActivity>(
            PaymentMethodsActivityStarter.Args.Builder()
                .setInitialPaymentMethodId(paymentMethod.id)
                .setPaymentConfiguration(PaymentConfiguration.getInstance(context))
                .build()
        ).use { activityScenario ->
            activityScenario.onActivity {
                val recyclerView = it.viewBinding.recycler

                verify(customerSession)
                    .getPaymentMethods(
                        eq(PaymentMethod.Type.Card),
                        isNull(),
                        isNull(),
                        isNull(),
                        eq(setOf(PaymentMethodsActivity.PRODUCT_TOKEN)),
                        listenerArgumentCaptor.capture()
                    )

                listenerArgumentCaptor.firstValue
                    .onPaymentMethodsRetrieved(PaymentMethodFixtures.CARD_PAYMENT_METHODS)

                val paymentMethodsAdapter =
                    recyclerView.adapter as PaymentMethodsAdapter
                assertEquals(paymentMethod.id, paymentMethodsAdapter.selectedPaymentMethod?.id)
            }
        }
    }

    @Test
    fun onClickAddClickView_withoutPaymentSession_launchesWithoutLog() {
        activityScenarioFactory.create<PaymentMethodsActivity>(
            PaymentMethodsActivityStarter.Args.Builder()
                .setPaymentConfiguration(PaymentConfiguration.getInstance(context))
                .build()
        ).use { activityScenario ->
            activityScenario.onActivity { activity ->
                val addCardView: View = activity
                    .findViewById(R.id.stripe_payment_methods_add_card)

                addCardView.performClick()

                val intentForResult =
                    shadowOf(activity).nextStartedActivityForResult
                val component = intentForResult.intent.component
                assertEquals(AddPaymentMethodActivity::class.java.name, component?.className)
            }
        }
    }

    @Test
    fun onClickAddCardView_whenStartedFromPaymentSession_launchesActivityWithLog() {
        activityScenarioFactory.create<PaymentMethodsActivity>(
            PaymentMethodsActivityStarter.Args.Builder()
                .setPaymentConfiguration(PaymentConfiguration.getInstance(context))
                .setIsPaymentSessionActive(true)
                .build()
        ).use { activityScenario ->
            activityScenario.onActivity { activity ->
                val addCardView: View = activity
                    .findViewById(R.id.stripe_payment_methods_add_card)
                addCardView.performClick()
                val intentForResult = shadowOf(activity).nextStartedActivityForResult
                val component = intentForResult.intent.component
                assertEquals(AddPaymentMethodActivity::class.java.name, component?.className)
                assertTrue(
                    AddPaymentMethodActivityStarter.Args.create(intentForResult.intent)
                        .isPaymentSessionActive
                )
            }
        }
    }

    @Test
    fun onActivityResult_withValidPaymentMethod_refreshesPaymentMethods() {
        activityScenarioFactory.create<PaymentMethodsActivity>(
            PaymentMethodsActivityStarter.Args.Builder()
                .setPaymentConfiguration(PaymentConfiguration.getInstance(context))
                .build()
        ).use { activityScenario ->
            activityScenario.onActivity { activity ->
                val progressBar = activity.viewBinding.progressBar
                val recyclerView = activity.viewBinding.recycler

                val paymentMethod = PaymentMethodFixtures.CARD_PAYMENT_METHODS[2]
                assertNotNull(paymentMethod)

                val resultIntent = Intent()
                    .putExtras(
                        AddPaymentMethodActivityStarter.Result.Success(paymentMethod)
                            .toBundle()
                    )
                activity.onActivityResult(
                    AddPaymentMethodActivityStarter.REQUEST_CODE,
                    RESULT_OK,
                    resultIntent
                )
                assertEquals(View.VISIBLE, progressBar.visibility)
                verify(customerSession, times(2))
                    .getPaymentMethods(
                        eq(PaymentMethod.Type.Card),
                        isNull(),
                        isNull(),
                        isNull(),
                        eq(setOf(PaymentMethodsActivity.PRODUCT_TOKEN)),
                        listenerArgumentCaptor.capture()
                    )

                listenerArgumentCaptor.firstValue
                    .onPaymentMethodsRetrieved(PaymentMethodFixtures.CARD_PAYMENT_METHODS)
                assertEquals(View.GONE, progressBar.visibility)
                assertEquals(4, recyclerView.adapter?.itemCount)

                val paymentMethodsAdapter =
                    recyclerView.adapter as PaymentMethodsAdapter

                paymentMethodsAdapter.selectedPaymentMethodId = paymentMethod.id
                assertEquals(paymentMethod.id, paymentMethodsAdapter.selectedPaymentMethod?.id)
            }
        }
    }

    @Test
    fun setSelectionAndFinish_finishedWithExpectedResult() {
        activityScenarioFactory.create<PaymentMethodsActivity>(
            PaymentMethodsActivityStarter.Args.Builder()
                .setPaymentConfiguration(PaymentConfiguration.getInstance(context))
                .build()
        ).use { activityScenario ->
            activityScenario.onActivity { activity ->
                val progressBar = activity.viewBinding.progressBar
                val recyclerView = activity.viewBinding.recycler
                val addCardView: View = activity.findViewById(R.id.stripe_payment_methods_add_card)

                verify(customerSession).getPaymentMethods(
                    eq(PaymentMethod.Type.Card),
                    isNull(),
                    isNull(),
                    isNull(),
                    eq(setOf(PaymentMethodsActivity.PRODUCT_TOKEN)),
                    listenerArgumentCaptor.capture()
                )

                assertEquals(View.VISIBLE, progressBar.visibility)
                assertEquals(View.VISIBLE, addCardView.visibility)
                assertEquals(View.VISIBLE, recyclerView.visibility)

                listenerArgumentCaptor.firstValue
                    .onPaymentMethodsRetrieved(PaymentMethodFixtures.CARD_PAYMENT_METHODS)
                val paymentMethodsAdapter =
                    recyclerView.adapter as PaymentMethodsAdapter
                paymentMethodsAdapter.selectedPaymentMethodId =
                    PaymentMethodFixtures.CARD_PAYMENT_METHODS[0].id

                activity.onBackPressed()

                // Now it should be gone.
                assertEquals(View.GONE, progressBar.visibility)
                assertTrue(activity.isFinishing)

                // `resultCode` is `RESULT_CANCELED` because back was pressed
                assertEquals(RESULT_CANCELED, activityScenario.result.resultCode)

                val result =
                    PaymentMethodsActivityStarter.Result.fromIntent(activityScenario.result.resultData)
                assertEquals(PaymentMethodFixtures.CARD_PAYMENT_METHODS[0], result?.paymentMethod)
            }
        }
    }
}
